### PR TITLE
fix: useScaffoldEventHistory caching

### LINKED ISSUE #903: useScaffoldEventHistory uses data from old fetch with different filters
### Is there an existing issue for this?

- [X] I have looked through the [existing issues](https://github.com/scaffold-eth/scaffold-eth-2/issues)

### Which method was used to setup Scaffold-ETH 2 ?

git clone

### Current Behavior

I use useScaffoldEventHistory twice on a single page with different filters. It fetches right data. I run again and the old data does not change, it use overrides the data with first run of useScaffoldEventHistory (even the constants have a differnt name). Any ideas??

This:

```
<PostCardList key="-1" filters={{ _post_hash: postHash }} />
<PostCardList key="-2" filters={{ _parent_post_hash: postHash }} />
```
and PostCardList.tsx:

```
export const PostCardList = ({ filters }: PostCardListProps) => {

const { data: postCreatedEvents, isLoading } = useScaffoldEventHistory({
    contractName: "PostCreator",
    eventName: "PostCreated",
    fromBlock: 0n,
    watch: true,
    filters: filters 
  });
...
```

But the second One gets overridden with contents of first one?? Both solidity parameters are indexed. And I also put a key inside my looped list items (unique hash)

### Expected Behavior

_No response_

### Steps To Reproduce

Use useScaffoldEventHistory on same page with different filters

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by rin-st**
## Description

Fixes useScaffoldEventHistory caching

Fixes #903

## Additional Information

- [x] I have read the [contributing docs](/scaffold-eth/scaffold-eth-2/blob/main/CONTRIBUTING.md) (if this is your first contribution)
- [x] This is not a duplicate of any [existing pull request](https://github.com/scaffold-eth/scaffold-eth-2/pulls)

### CODE REVIEW

**Review by technophile-04 — COMMENTED** [review]
Ohh I think the main underlying problem was I missed adding `filters` in `queryKey` : 

https://github.com/scaffold-eth/scaffold-eth-2/blob/a132cce62c1064bacff8dc1722115bd9200153f7/packages/nextjs/hooks/scaffold-eth/useScaffoldEventHistory.ts#L100-L108

Since we didn't had `filters` in query key `react-query` was treating both queries as same and hence not respecting the filters provided.

**Comment by technophile-04** [comment]
Just pushed commits to updated here is the demo fixing [mrr](https://github.com/phaitonican/scuffold-eth-2-event-history-test) mentioned [here](https://github.com/scaffold-eth/scaffold-eth-2/issues/903#issuecomment-2285574189)


https://github.com/user-attachments/assets/62a5b9fc-8b60-4664-9d9f-8253761db43b

File: packages/nextjs/hooks/scaffold-eth/useScaffoldEventHistory.ts [hunk 1]
```diff
@@ -105,6 +106,7 @@ export const useScaffoldEventHistory = <
         eventName,
         fromBlock: fromBlock.toString(),
         chainId: targetNetwork.id,
+        filters: JSON.stringify(filters, replacer),
```
Comment (technophile-04) [inline packages/nextjs/hooks/scaffold-eth/useScaffoldEventHistory.ts:109]: We could have passed the `filters` obj directly but sometimes there can be bigInt init it too. When react-query make a queryHash of all this items it uses JSON.stringify internally and not able to parse bigInt. 

So here we are making the react-query's life easier by handling the object with bigInt earlier so its easier for it to create query hash

**Comment by rin-st** [comment]
Can't approve my own PR 😄 . Thank you for update, lgtm!

**Review by technophile-04 — APPROVED** [review]
Merging this 🙌

**Pull request merged**